src/compiler/checker.ts

```diff
@@ -27281,16 +27281,29 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return false;
     }
 
-    function isDiscriminantProperty(type: Type | undefined, name: __String) {
+    // The flag `considerNonUniformPrimitivePropDiscriminant`, when enabled, introduces a new criterion for a property to be considered discriminant:
+    // 1 The property must be non-uniform
+    // 2 The property must include at least one primitive type as a possible type
+    function isDiscriminantProperty(type: Type | undefined, name: __String, considerNonUniformPrimitivePropDiscriminant: boolean = true) {
         if (type && type.flags & TypeFlags.Union) {
             const prop = getUnionOrIntersectionProperty(type as UnionType, name);
             if (prop && getCheckFlags(prop) & CheckFlags.SyntheticProperty) {
+                const propType = getTypeOfSymbol(prop);
                 // NOTE: cast to TransientSymbol should be safe because only TransientSymbols can have CheckFlags.SyntheticProperty
-                if ((prop as TransientSymbol).links.isDiscriminantProperty === undefined) {
-                    (prop as TransientSymbol).links.isDiscriminantProperty = ((prop as TransientSymbol).links.checkFlags & CheckFlags.Discriminant) === CheckFlags.Discriminant &&
-                        !isGenericType(getTypeOfSymbol(prop));
+                const transientSymbol = prop as TransientSymbol;
+                transientSymbol.links.isDiscriminantProperty ??= new Map();
+
+                let isDiscriminant = transientSymbol.links.isDiscriminantProperty.get(considerNonUniformPrimitivePropDiscriminant);
+
+                if (typeof isDiscriminant === "undefined") {
+                    isDiscriminant = (((transientSymbol.links.checkFlags & CheckFlags.Discriminant) === CheckFlags.Discriminant)
+                        || !!(considerNonUniformPrimitivePropDiscriminant && (transientSymbol.links.checkFlags & CheckFlags.HasNonUniformType) && someType(propType, t => !!(t.flags & TypeFlags.Primitive))))
+                        && !isGenericType(propType);
+

```

Comment:
This check can probably be done in `createUnionOrIntersectionProperty`, setting a new `CheckFlag.HasPrimitive`, like we do for `HasNonUniformType`.

Reply:
Even if it doesn't matter for the perf benchmarks we have, I think yes, to be consistent and keep similar code in a single place.

Reply:
@gabritto some feedback.

The basic change is quite easy, but there are some problems because `createUnionOrIntersectionProperty` terminates with `result.links.type = isUnion ? getUnionType(propTypes) : getIntersectionType(propTypes)`, and is this resulting type the one on which I'm currently running `someType(propType, t => !!(t.flags & TypeFlags.Primitive)))`. It's like it has already been reduced/cleaned up/whatever.

Just setting to `1` the `CheckFlag.HasPrimitiveType` beforehand like we do for other flags doesn't seem a good idea. I got 8 failing baselines and some changes are pretty weird. I'd suspect the right condition for setting the `HasPrimitiveType` flag will be a little convoluted.

I pushed the current changes so you can have a look and maybe spot some stupid error I'm doing, or provide some insight for the above condition.

Reply:
@gabritto plot twist, I tried to disable the new discriminating flag everywhere except in the only place I was sure we need it and things have improved a lot. The new discriminating power is being activated in more cases with your suggestion, that's why I changed more baselines, please check them.

Reply:
Pushed a fix that you may dislike, so I'm open to suggestions. 

The problem:

```ts
node.body // has type body?: ModuleBody | undefined
```

The condition `node.body` is checked so its type is refined to just `ModuleBody`. That's fine!

Then you check `!node.body.parent` where:

```ts
node.body.parent // has type ModuleDeclaration | Node | ModuleBody | SourceFile
```

It cannot be non defined! So the check `!node.body.parent` now narrows the type of `node.body` from `ModuleBody` to `never`.